service/accessanalyzer/paginators.go

```diff
@@ -0,0 +1,307 @@
+// Code generated by smithy-go-codegen DO NOT EDIT.
+
+package accessanalyzer
+
+import (
+	"context"
+	"fmt"
+)
+
+// ListFindingsAPIClient is a client that implements the ListFindings operation.
+type ListFindingsAPIClient interface {
+	ListFindings(context.Context, *ListFindingsInput, ...func(*Options)) (*ListFindingsOutput, error)
+}
+
+var _ ListFindingsAPIClient = (*Client)(nil)
+
+// ListFindingsPaginatorOptions is the paginator options for ListFindings
+type ListFindingsPaginatorOptions struct {
+	// The maximum number of results to return in the response.
+	Limit *int32
+
+	// Set to true if pagination should stop if the service returns the same taken
+	// passed to it
+	StopOnDuplicateToken bool
+}
+
+// ListFindingsPaginator is a paginator for
+// github.com/aws/aws-sdk-go-v2/service/accessanalyzer.ListFindings
+type ListFindingsPaginator struct {
+	options   ListFindingsPaginatorOptions
+	client    ListFindingsAPIClient
+	params    *ListFindingsInput
+	nextToken *string
+	firstPage bool
+	done      bool
+}
+
+// NewListFindingsPaginator returns a new ListFindingsPaginator
+func NewListFindingsPaginator(client ListFindingsAPIClient, params *ListFindingsInput, optFns ...func(*ListFindingsPaginatorOptions)) *ListFindingsPaginator {
+	options := ListFindingsPaginatorOptions{}
+	for _, fn := range optFns {
+		fn(&options)
+	}
+	return &ListFindingsPaginator{
+		options:   options,
+		client:    client,
+		params:    params,
+		firstPage: true,
+	}
+}
+
+// HasMorePages returns a boolean indicating whether more pages are available
+func (p *ListFindingsPaginator) HasMorePages() bool {
+	if p.done {
+		return false
+	}
+	return p.firstPage || (p.nextToken != nil && len(*p.nextToken) > 0)
+}
+
+// NextPage retrieves the next ListFindings page.
+func (p *ListFindingsPaginator) NextPage(ctx context.Context, optFns ...func(*Options)) (*ListFindingsOutput, error) {
+	if !p.HasMorePages() {
+		return nil, fmt.Errorf("no more pages available")
+	}
+
+	params := *p.params
+	params.NextToken = p.nextToken
+	params.MaxResults = p.options.Limit
+	result, err := p.client.ListFindings(ctx, &params, optFns...)
+	if err != nil {
+		return nil, err
+	}
+	p.firstPage = false
+	p.nextToken = result.NextToken
+	if p.options.StopOnDuplicateToken && params.NextToken != nil && p.nextToken != nil && *params.NextToken == *p.nextToken {
+		p.done = true
+	} else {
+		p.done = false
+	}
```

Comment:
With the terminal error case above `if !p.HasMorePages` is the `else` case here possible?  If using `nextToken` of `nil` in the case of duplicate tokens would replicate the behavior of service returning no next token, and would remove the need for the `done` member.